language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/LangExtensionDelegator.java

```diff
@@ -0,0 +1,110 @@
+import org.eclipse.lsp4j.CompletionList;
+import org.eclipse.lsp4j.CompletionParams;
+import org.eclipse.lsp4j.DocumentFormattingParams;
+import org.eclipse.lsp4j.PublishDiagnosticsParams;
+import org.eclipse.lsp4j.TextEdit;
+import org.eclipse.lsp4j.jsonrpc.messages.Either;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.ServiceLoader;
+
+/**
+ * Delegate the operation to the valid extension.
+ *
+ * @since 2.0.0
+ */
+public class LangExtensionDelegator {
+
+    private static final LangExtensionDelegator INSTANCE = new LangExtensionDelegator();
+
+    private final List<CompletionExtension> completionExtensions = new ArrayList<>();
+    private final List<FormattingExtension> formatExtensions = new ArrayList<>();
+    private final List<DiagnosticsExtension> diagExtensions = new ArrayList<>();
+
+    private LangExtensionDelegator() {
+        ServiceLoader.load(LanguageExtension.class).forEach(languageExtension -> {
+            switch (languageExtension.kind()) {
+                case COMPLETION:
+                    completionExtensions.add((CompletionExtension) languageExtension);
+                    break;
+                case FORMAT:
+                    formatExtensions.add((FormattingExtension) languageExtension);
+                    break;
+                case DIAGNOSTIC:
+                    diagExtensions.add((DiagnosticsExtension) languageExtension);
+                    break;
+                default:
+                    break;
+            }
+        });
+    }
+
+    /**
+     * Get the completions.
+     *
+     * @param params completion parameters
+     * @return {@link Either} completion results
+     */
+    public Either<List<CompletionItem>, CompletionList> completion(CompletionParams params, LSContext context) {
+        return completionExtensions.stream().filter(ext -> ext.validate(params))
+                .map(ext -> ext.execute(params, context))
+                .findFirst()
+                .orElse(Either.forRight(new CompletionList()));
+    }
+
+    /**
+     * Get the formatting.
+     *
+     * @param params formatting parameters
+     * @return {@link List} of text edits
+     */
+    public List<? extends TextEdit> formatting(DocumentFormattingParams params, LSContext context) {
+        return formatExtensions.stream().filter(ext -> ext.validate(params))
+                .map(ext -> ext.execute(params, context))
+                .findFirst()
+                .orElse(new ArrayList<>());
+    }
+
+    /**
+     * Get the Diagnostics.
+     *
+     * @param uri document URI
+     * @return {@link PublishDiagnosticsParams} diagnostic params calculated
+     */
+    public PublishDiagnosticsParams diagnostics(String uri, LSContext context) {
+        return diagExtensions.stream().filter(ext -> ext.validate(uri))
+                .map(ext -> ext.execute(uri, context))
+                .findFirst()
+                .orElse(new PublishDiagnosticsParams());
```

Comment:
Specific reason to limit to one? We can merge the results from multiple filtered extensions. eg. there can be two completion extensions with the same uri pattern.

Reply:
Yes, I added this one to force only one at this time because we need to decide whether we collect from all the extensions and if we do so, do we need a specific enable disable mechanism (Because as an example, for completions and signatures, ordering of the items matters). We need to decide that. Otherwise I also feel we should allow all and +1 for proceed.
@rasika Shall we allow all in the first phase and shall we decide later about the prioritisation and other similar capabilities?

Reply:
Added the aggregated result support